### PR TITLE
add more concise JSON report to flaky-test-reporter

### PR DESCRIPTION
**Pull request by TrevorFarrelly**
**What this PR does, why we need it**:
This is the first main step in the implementation of the flaky test re-trier. The re-trier needed a more concise logs of the currently flaky tests than what build artifacts can provide. 

This will generate a file named `all-flaky-tests.json` in the local artifacts directory, containing the set of all currently flaky tests for that repository.

/lint

**Which issue(s) this PR fixes**:
Fixes #

**Special notes to reviewers**:

**User-visible changes in this PR**:

### CODE REVIEW

**Comment by chaodaiG** [comment]
I suggest breaking this up into a separate package, so it can be used by both flaky-test-reporter and retryer. This package shouldn't need to understand anything about how flaky tests were generated, and shouldn't need to import anything from flaky-test-reporter

File: tools/flaky-test-reporter/json-report.go [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"path"
+
+	"github.com/knative/test-infra/shared/prow"
+)
+
+type JSONReport struct {
```
Comment (srinivashegde86) [inline tools/flaky-test-reporter/json-report.go]:
```suggestion
type Report struct {
```

File: tools/flaky-test-reporter/json-report.go [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+	Flaky []string `json:"flaky"`
+}
+
+func (jr *JSONReport) writeToArtifactsDir(repo string, dryrun bool) error {
+	artifactsDir := prow.GetLocalArtifactsDir()
+	// no need to create the directory, it is guaranteed to exist since artifacts were generated earlier
+	outFilePath := path.Join(artifactsDir, repo, "all-flaky-tests.json")
```
Comment (srinivashegde86) [inline tools/flaky-test-reporter/json-report.go]: is there a non-all option?
```suggestion
	outFilePath := path.Join(artifactsDir, repo, "flaky-tests.json")
```

File: tools/flaky-test-reporter/json-report.go [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+}
+
+// we want all flaky tests across all jobs for a given repo. There can be duplicate flaky tests
+// if multiple jobs use the same test cases. We remove duplicates by using a map as a "set".
+func getFlakyTestSet(repo string, repoDataAll []*RepoData) []string {
+	testSet := make(map[string]bool)
+	for _, rd := range repoDataAll {
+		if rd.Config.Repo == repo {
+			for _, test := range getFlakyTests(rd) {
+				testSet[test] = true
```
Comment (srinivashegde86) [inline tools/flaky-test-reporter/json-report.go]: you can append it in this loop itself and dont need another one

File: tools/flaky-test-reporter/json-report.go [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+	}
+	var flakyTests []string
+	for test := range testSet {
+		flakyTests = append(flakyTests, test)
+	}
+	return flakyTests
+}
+
+func writeFlakyTestsToJSON(repoDataAll []*RepoData, dryrun bool) error {
+	for repo := range jobConfigs {
```
Comment (srinivashegde86) [inline tools/flaky-test-reporter/json-report.go]: I am not sure if we need another method that just loops over jobConfigs and calls writeToArtifacts(). Probably, we can just combine them into one method.

File: tools/flaky-test-reporter/main.go [hunk 5]
```diff
@@ -77,13 +77,17 @@ func main() {
...
+	jsonErr := writeFlakyTestsToJson(repoDataAll, *dryrun)
 	if nil != githubErr {
 		log.Printf("Github step failures:\n%v", githubErr)
 	}
 	if nil != slackErr {
 		log.Printf("Slack step failures:\n%v", slackErr)
 	}
-	if nil != githubErr || nil != slackErr { // Fail this job if there is any error
+	if nil != jsonErr {
+		log.Printf("JSON step failures:\n%v", jsonErr)
```
Comment (srinivashegde86) [inline tools/flaky-test-reporter/main.go]: more meaningful msg? -> "Failed writing report to artifacts: %v"

**Comment by TrevorFarrelly** [comment]
Closing this PR, I will re-implement as a shared library.

**Pull request closed without merging**